File: src/pipeline/briefing.ts
import type { AnyDb } from '../db/index.js';
import { briefings } from '../db/schema.js';
import type { Category, Urgency, WorkItem } from '../models/types.js';
import { workItemLabel, workItemSummary } from '../models/types.js';
import type { StateDiff } from './state.js';

const SECTION_ORDER: Category[] = ['needs_review', 'needs_action', 'blocked', 'waiting_on_others', 'slack_mention', 'fyi'];
const SECTION_TITLES: Record<Category, string> = {
  needs_review: 'Needs your review',
  needs_action: 'Needs your action',
  blocked: 'Blocked / on hold',
  waiting_on_others: 'Waiting on others',
  slack_mention: 'Slack mentions (no linked ticket)',
  fyi: 'FYI',
};
const URGENCY_RANK: Record<Urgency, number> = { high: 0, medium: 1, low: 2 };

export function generateBriefing(items: WorkItem[], diff: StateDiff, runDate: string): string {
  const lines: string[] = [`# Morning Briefing — ${runDate}`, ''];
  lines.push(
    `Tracking **${items.length}** work items (${diff.newItems.length} new, ${diff.changed.length} changed, ${diff.resolved.length} resolved since last run).`,
  );
  lines.push('');

  const byCategory = new Map<Category, WorkItem[]>();
  for (const item of items) {
    const list = byCategory.get(item.category) ?? [];
    list.push(item);
    byCategory.set(item.category, list);
  }

  for (const category of SECTION_ORDER) {
    const catItems = byCategory.get(category);
    if (!catItems || catItems.length === 0) continue;
    catItems.sort((a, b) => URGENCY_RANK[a.urgency] - URGENCY_RANK[b.urgency]);
    lines.push(`## ${SECTION_TITLES[category]} (${catItems.length})`);
    for (const item of catItems) {
      const flags: string[] = [];
      if (diff.newItems.includes(item.id)) flags.push('NEW');
      if (diff.changed.includes(item.id)) flags.push('CHANGED');
      const flagStr = flags.length ? ` [${flags.join(', ')}]` : '';
      lines.push(`- **${workItemLabel(item)}** (${item.urgency})${flagStr} — ${workItemSummary(item)}`);
      for (const reason of item.reasons) lines.push(`  - ${reason}`);
    }
    lines.push('');
  }

  if (diff.resolved.length > 0) {
    lines.push(`## Resolved since last run (${diff.resolved.length})`);
    for (const id of diff.resolved) lines.push(`- ${id}`);
    lines.push('');
  }

  return lines.join('\n').replace(/\n+$/, '') + '\n';
}

export async function persistBriefing(db: AnyDb, runDate: string, content: string, kind = 'daily'): Promise<void> {
  await db
    .insert(briefings)
    .values({ kind, runDate, content, createdAt: new Date() })
    .onConflictDoUpdate({ target: [briefings.kind, briefings.runDate], set: { content, createdAt: new Date() } });
}

File: legacy-python/work_agent/briefing.py
from __future__ import annotations

from collections import defaultdict
from typing import List

from .models import WorkItem
from .state import StateDiff

_SECTION_ORDER = ["needs_review", "needs_action", "blocked", "waiting_on_others", "slack_mention", "fyi"]
_SECTION_TITLES = {
    "needs_review": "Needs your review",
    "needs_action": "Needs your action",
    "blocked": "Blocked / on hold",
    "waiting_on_others": "Waiting on others",
    "slack_mention": "Slack mentions (no linked ticket)",
    "fyi": "FYI",
}
_URGENCY_RANK = {"high": 0, "medium": 1, "low": 2}


def generate_briefing(items: List[WorkItem], diff: StateDiff, run_date: str) -> str:
    lines = [f"# Morning Briefing — {run_date}", ""]
    lines.append(
        f"Tracking **{len(items)}** work items "
        f"({len(diff.new)} new, {len(diff.changed)} changed, {len(diff.resolved)} resolved since last run)."
    )
    lines.append("")

    by_category: dict[str, list[WorkItem]] = defaultdict(list)
    for item in items:
        by_category[item.category].append(item)

    for category in _SECTION_ORDER:
        cat_items = by_category.get(category, [])
        if not cat_items:
            continue
        cat_items = sorted(cat_items, key=lambda i: _URGENCY_RANK.get(i.urgency, 9))
        lines.append(f"## {_SECTION_TITLES[category]} ({len(cat_items)})")
        for item in cat_items:
            flags = []
            if item.id in diff.new:
                flags.append("NEW")
            if item.id in diff.changed:
                flags.append("CHANGED")
            flag_str = f" [{', '.join(flags)}]" if flags else ""
            lines.append(f"- **{item.label()}** ({item.urgency}){flag_str} — {item.summary()}")
            for reason in item.reasons:
                lines.append(f"  - {reason}")
        lines.append("")

    if diff.resolved:
        lines.append(f"## Resolved since last run ({len(diff.resolved)})")
        for item_id in diff.resolved:
            lines.append(f"- {item_id}")
        lines.append("")

    return "\n".join(lines).rstrip() + "\n"
